Upgrades to CoordinateSystem and graphing.

hydrobeam opened the pull request:
<!--
Thanks for your contribution to ManimCommunity!

Before filling in the details, ensure:
- Your local changes are up-to-date with ManimCommunity/manim
  
- The title of your PR gives a descriptive summary to end-users. Some examples:
  - Fixed last animations not running to completion
  - Added gradient support and documentation for SVG files
  Examples of what *NOT* to do:
  - "fixed that styling issue" - not descriptive enough
  - "fixed issue #XYZ" - end-user needs to do further research
-->

## Changelog/Overview

<!--changelog-start-->
Breaking changes were introduced to :class:`Axes`, :class:`ThreeDAxes`, :class:`NumberPlane` and :class:`NumberLine`
All the above now use lists to construct their ranges as opposed to explicitly defining these values --> `x_range`  :  `[x_min, x_max, x_step]`. There were also many upgrades to these classes which improves their functionality and ease of use.

Furthermore, `~.GraphScene` has been deprecated and we encourage use of the above functions instead. It can still be used because :class:`NumberLineOld` was introduced to accommodate these changes.
<!--changelog-end-->

## Motivation
<!-- Outline your motivation: In what way do your changes improve the library? -->
The current handling of graphing logic with `coordinate_systems.py` was outdated and not very intuitive. This PR ports over the changes made upstream to `coordinate_systems.py`, `numberline.py` and `functions.py` which should make the process of graphing less convoluted. 

## Overview / Explanation for Changes
### `coordinate_systems.py` 
- Rework `CoordinateSystem` to accommodate `x_range` and `y_range` and their corresponding length values.
   - Ranges represent ``[min, max, step]``
   - `get_graph` updated to support the range.
   - `get_line_from_axis_to_point` introduced.
- Rework `Axes`
  - Uses `x_range` and `y_range` to create the axes.
  - default `x_length` and `y_length` changed so that the axes are each two units away from the edges of the screen (see below)
![image](https://user-images.githubusercontent.com/70682032/112204021-90eca200-8be9-11eb-964b-bd0ac0ccf06a.png)
   - `create_axis` automatically shifts the axes based on the zero point (if it does not exist, shifts to the max/min)
   - `stroke_width` reduced to `2` resulting in thinner axes.
   - `ThreeDAxes` compatibility with Axes.
- `NumberPlane` compatiblity with Axes
   - `NumberPlane`, is now enclosed by lines(box shaped) and is smaller.
   - The plane adjusts  its size based on the `x_range` and `y_range` value so that the unit size is always one (represents coordinates).
   - New
![image](https://user-images.githubusercontent.com/70682032/115942800-ef6f9f00-a479-11eb-916c-6fc077e5eae1.png)
   - Old
![image](https://user-images.githubusercontent.com/70682032/112203748-38b5a000-8be9-11eb-8090-4038d1667a05.png)


- `ComplexPlane` support 
 
### `number_line.py`
- `x_min`, `x_max`, `add_start`, `add_end`, `tick_frequency`, `leftmost_tick`, `number_at_center`, `numbers_to_show` (might re-introduce `numbers_to_show`) removed
- `numbers_to_exclude` replaces `exclude_zero_from_default_numbers`, `length` replaces `width`
- `stroke_width` and `x_range` added.
-  Redundant methods removed, some methods renamed to be clearer, other methods reworked/updated. (such as `add_numbers`)
- `UnitInterval` updated to reflect changes.

### `functions.py`
- Accommodate using ranges.

### `graph_scene.py`
- Deprecated.
- OldNumberLine created in `number_line.py` to accommodate these changes.

Code to recreate default GraphScene axes:
```python
class AxesShowcase(Scene):
    def construct(self):
        graph = Axes(
            x_range=[-1,10,1],
            y_range=[-1,10,1],
            x_length=9,
            y_length=6,
            axis_config={"include_tip":False}
        )
        labels = graph.get_axis_labels()
        self.add(graph, labels)
```
![image](https://user-images.githubusercontent.com/70682032/112203896-68fd3e80-8be9-11eb-9a33-6be809921128.png)


## Testing Status
<!-- Optional (but recommended): your computer specs and
what tests you ran with their results, if any. This section
is also intended for other testing-related comments. -->

I modified some tests to accommodate the changes. This included removing the usage of `GraphScene` for some tests and using `Axes` for some tests, but generally they should all produce the same output, or at least similar. The test that was entirely reworked was the NumberPlane test since it involved splitting `x_min` and `x_max` between the `axis_config`s and attributes, which is not possible with the new range system.

## Further Comments
<!-- Optional, any further comments regarding your PR
that might be useful for reviewers.. -->



A bit of the code that was added is currently redundant (see: `font_size` in `NumberLine` for example) because it relies on updated functionality from other files. I'm not sure what the stance is on updating those, so I left it alone. 

It is worth noting that trying to recreate some of the code in the documentation and tests using solely `Axes` was at times janky, so I might also port over some tools used by `GraphScene` into `Axes`. Although, the code was much cleaner and simpler for most tasks.
 
I will be adding the documentation changes in the next PR.

## Acknowledgements
- [x] I have read the [Contributing Guidelines](https://docs.manim.community/en/latest/contributing.html)
- [x] I have chosen a descriptive PR title (see top of PR template for examples)
<!-- Once again, thanks for helping out by contributing to manim! -->


<!-- Do not modify the lines below. -->
## Reviewer Checklist
- [ ] Newly added functions/classes are either private or have a docstring
- [ ] Newly added functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [ ] Newly added documentation builds, looks correctly formatted, and adds no additional build warnings
- [ ] The PR title is descriptive enough

GameDungeon commented:
Could you re-write the test you are now failing? These are some nice improvements and I'm quite happy you have made this.

GameDungeon commented on manim/mobject/coordinate_systems.py:
```diff
@@ -5,19 +5,19 @@
 from ..constants import *
 from ..mobject.functions import ParametricFunction
-from ..mobject.geometry import Arrow
+from ..mobject.geometry import Arrow, DashedLine
 from ..mobject.geometry import Line
```
Nitpicky, but could you make the imports in line pelase.
```suggestion
from ..mobject.geometry import Arrow, DashedLine, Line
```

GameDungeon reviewed (approved):
LGTM

Needs typing but that should be added in a later pull. Could this be merged quickly, as it is extremely useful? I can't overstate how happy I am with the changes added here.

kolibril13 commented on manim/mobject/coordinate_systems.py:
```diff
@@ -87,14 +120,33 @@ def get_axis_labels(self, x_label_tex="x", y_label_tex="y"):
         )
         return self.axis_labels
 
+    def get_line_from_axis_to_point(
```
is this not the same like `get_vertical_line_to_graph`?
In case that not, maybe you can add an example here:
https://manimce--1161.org.readthedocs.build/en/1161/examples.html#sinandcosfunctionplot
so one can see in action what the difference between these two is

hydrobeam replied:
I compared the two and I think they're a bit different: `axis_to_point` takes in a point anywhere on the graph, whereas `line_to_graph` is specifically linked to a predefined function. They're definitely at least similar. That example is already a little busy, so adding another line doesn't seem like the best idea though.

kolibril13 replied:
Okay, nice, that makes sense.
Then we can e.g. add it here:
https://manimce--1161.org.readthedocs.build/en/1161/examples.html#graphareaplot

hydrobeam replied:
I just realized that it actually won't work because this method is in `CoordinateSystem` and `GraphScene` does not rely on `CoordinateSystem` at all. The documentation changes should probably be saved when I port the examples to `Axes`.

kolibril13 commented:
> The next step is to shift over `GraphScene`'s useful methods over to `CoordinateSystem` and deprecate `GraphScene`.

Great idea.
Thanks for this amazing addition.
One thing I just realized when looking over the old and the new docs is that the size of the axis decreased:
![image](https://user-images.githubusercontent.com/44469195/112732862-db995180-8f3c-11eb-8b0c-3158a788f398.png)
Is that intentional?

hydrobeam commented:
For some reason I can't reply to the previous comment 😅. But yes the `stroke_width` change is intentional. I noticed that the upstream version used thinner axes and I was a fine of the style; but it's very easily reversible. I'm open to hearing opinions about it!

kolibril13 commented:
I have not too strong feelings about the `stroke_width`, we can keep it like it is.
Is it possible to fine-tune the `stroke_width` for the axes with a user facing parameter?

hydrobeam commented:
Yep, there's a `stroke_width` parameter included in `NumberLine` that defaults to 2.0.

kolibril13 commented:
I suggest that we make then one of the examples in the example gallery with thicker stroke width and choose the thin stroke width as default.

GameDungeon commented:
Here is an idea: Make X and Y Length a ratio. the current 9 & 6 being [3, 2]. It simplifies the args, might want others opinions on this as it might convolute things.

behackl reviewed (commented):
Thank you for your efforts! I think this is already a pretty good direction, however there are some issues I'd like to have discussed before moving this further. I've left a bunch of suggestions and comments that are up for discussion.

GameDungeon commented on manim/scene/graph_scene.py:
```diff
@@ -77,15 +75,13 @@ def construct(self):
 class GraphScene(Scene):
```
Honestly just revert the changes and deprecate it.

GameDungeon commented:
If everyone agrees with me, revert the changes to GraphScene and instead just deprecate it.

hydrobeam commented on manim/mobject/coordinate_systems.py:
```diff
@@ -128,56 +180,56 @@ def input_to_graph_point(self, x, graph):
-            "label_direction": LEFT,
-            "rotation": 90 * DEGREES,
-        }
+        self.axis_config = {"include_tip": True, "numbers_to_exclude": [0]}
```
Yup, it's set by default in `NumberLine` so it was redundant in this `axis_config`

hydrobeam commented on manim/mobject/coordinate_systems.py:
```diff
@@ -275,24 +340,53 @@ def make_func(axis):
     ):
+
+        # use ceilings and floors to make the box bound the screen when None is passed.
+        y_range = (
+            np.array(
+                [
+                    math.ceil(-config["frame_y_radius"]),
+                    math.floor(config["frame_y_radius"]),
+                    1.0,
+                ]
+            )
+            if y_range is None
+            else y_range
+        )
+
+        x_range = (
+            np.array(
+                [
+                    math.ceil(-config["frame_x_radius"]),
+                    math.floor(config["frame_x_radius"]),
+                    1.0,
+                ]
+            )
+            if x_range is None
+            else x_range
+        )
```
Using `super().__init__` would cass the number plane to inherit the axes currently reduced length and not fill up the entire screen. That's why i went with the approach of re-calling it and using the entire screen for `NumberPlane`

hydrobeam commented on manim/mobject/coordinate_systems.py:
```diff
@@ -467,8 +559,8 @@ def p2n(self, point):
     def get_default_coordinate_values(self):
-        x_numbers = self.get_x_axis().default_numbers_to_display()
-        y_numbers = self.get_y_axis().default_numbers_to_display()
+        x_numbers = self.get_x_axis().get_tick_range()[1:]
+        y_numbers = self.get_y_axis().get_tick_range()[1:]
```
Not sure, I tried it with and without and it just seems to cut off one of the edge numbers:

With:
![image](https://user-images.githubusercontent.com/70682032/113495895-a05dcc00-94c2-11eb-9f68-d9659a62ae4d.png)

Without: 
![image](https://user-images.githubusercontent.com/70682032/113495901-a81d7080-94c2-11eb-9737-f1a5b348ddff.png)

I guess it's to keep the sides balanced in terms of numbers but poorly executed? Or maybe just a relic.

hydrobeam commented:
I'll take your recommendations and create a `OldNumberLine` class with a deprecation warning. I'll revert the changes to `GraphScene` and make it so that it automatically calls this class.

GameDungeon commented:
Could this be updated to the latest version?

hydrobeam commented:
I've carried out the changes. `GraphScene` and its related tests/docs have been reverted to their original state and deprecated. To support `GraphScene`, I created a new class called `OldNumberLine` in `number_line.py` which uses the old system. Unfortunately, there are some conflicts and I'm not sure how to resolve them.

hydrobeam commented:
After some effort, I patched up the merge conflicts. It should be good to go!

GameDungeon commented:
Thanks so much :) What do you think @behackl now? It LGTM to me.

behackl commented on manim/mobject/geometry.py:
```diff
@@ -838,6 +838,14 @@ def get_unit_vector(self):
         return angle_of_vector(self.get_vector())
 
+    def get_projection(self, point):
+        """
+        Return projection of a point onto the line
```
Added punctuation. Also, a test would be nice -- either as a doctest (like in the suggestion), or as a separate (non-graphical) unit test in `tests`.

```suggestion
        """Return the projection of a point onto the line.

        Examples
        --------

        ::

            >>> import numpy as np
            >>> line = Line(LEFT, RIGHT)
            >>> line.get_projection(np.array([0, 1, 0]))
            array([0, 0, 0])

```

behackl commented on manim/mobject/number_line.py:
```diff
@@ -21,111 +22,103 @@
 class NumberLine(Line):
```
The way you've changed things right now still introduces a major breaking change to `NumberPlane`, and people will have to adapt their code to use `NumberPlaneOld` on their own -- it's the opposite direction of what I suggested above.

I won't block the PR over it, but I would certainly like to hear the opinions of other community devs before proceeding with this solution.

hydrobeam replied:
Even though both `GraphScene` and `NumberPlane` are widely used, switching code over from `GraphScene` to `Axes` requires significant effort, so I felt that creating a separate `OldNumberLine` was reasonable. The changes to `NumberPlane` *are* breaking, but I think that it's comparably very easy to adjust the code to get it back up and running. That's why I'm a little hesitant about adding a separate `NumberPlaneOld` class.

behackl replied:
It's fine, we can stick with introducing a breaking change to `NumberPlane`. In this case please:

1. update the PR description: within the 
   ```
   <!--changelog-start-->
   <!--changelog-end-->
   ```
   block, add a short migration guide from the old to the new usage of `NumberPlane`, and remark that the old mobject is still available via `NumberPlaneOld`.
2. Maybe also add a `logger.info` making users aware of the fact that this is the new interface and that the old class is available as `NumberPlaneOld`.

hydrobeam replied:
There actually isn't a `NumberPlaneOld` at the moment, just a `NumberLineOld` to work with `GraphScene`. Would you like me to add a `NumberPlaneOld`?

behackl replied:
No, I meant NumberLine / NumberLineOld.

behackl reviewed (commented):
Two more comments.

Due to the size of this PR it's unfortunately a bit difficult to review properly, but I more or less agree with the current status of the refactor. (It's definitely an improvement over the current situation.)

Before moving this forward, I'd like to get some additional opinions by other community devs as well. Thank you for your efforts!

GameDungeon commented:
Going off what behackl said change the old numberplane back to numberplane. Deprecate it. And then add numberplanenew.

Darylgolden commented:
Since we're deprecating `GraphScene`, we should definitely be replacing the examples in the docs that use it. It doesn't have to be done in this PR, but it should be done before the next release (it won't be nice to have 4 examples in our Example gallery made with a deprecated class). Thank you very much for the contribution, I'll be looking forward to using it!

hydrobeam commented:
I can take care of recreating existing scenes in the docs with `Axes` instead in a new PR, it'd just be impossible to review properly if I added them here😅. Plus, it'd be a good opportunity to try and see what Axes is missing and could use some work copying over from `GraphScene`.

GameDungeon commented:
Is this good now then?

abhi3019 commented on manim/mobject/coordinate_systems.py:
```diff
@@ -127,56 +171,55 @@ def input_to_graph_point(self, x, graph):
-        y_max=None,
+        x_range=None,
+        y_range=None,
+        x_length=config.frame_width - 2,
+        y_length=config.frame_height - 2,
```
In my opinion these lengths are really unnecessary because :
1. they seem very random
2. they can be passed via `x/y_axis_config = {"length": <length>}` or `axis_config = {"length": <length>}`
3. `unit_size` of the `x_axis` and `y_axis` is no longer respected (became redundant) because both `length` and `unit_size` shouldn't be passed into `NumberLine`. Try this code for example:
```py
plane = NumberPlane(
            x_range=[-2, 4],
            y_range=[-2, 3],
            axis_config={"include_numbers": True, "unit_size": 0.5},
        )
# print([axis.unit_size for axis in plane.axes]) --> [2.3703703703703702, 1.6]
self.add(plane)
```
No matter what value you pass for `unit_size` via `axis_config` or `x/y_axis_config`, you'll always get the same result.
For the above reasons, I'd suggest to remove the length parameters

hydrobeam replied:
1. The lengths are specified so that the axes don't take up the entire screen to make them look more visually appealing, but I'm not sure what you mean by random.
2. I think the length of the axes is a very important property to be able to easily modify. Working with axis configs is generally more tedious as opposed to an explicit parameter. There *is* a little redundancy, but I find the current implementation more intuitive.
3. I hadn't noticed this redundancy, but this makes me think that `unit_size` should be removed, not `x_length` and `y_length`.

abhi3019 replied:
>I'm not sure what you mean by random

I just meant that the `-2` part felt random, as in it could've been any other number. But we have to use some default length anyway, so that's fine



> 3\. this makes me think that `unit_size` should be removed, not `x_length` and `y_length`

None of those should be compromised in favor of the other imo. Both are extremely useful and removing one of them doesn't look good to me. Maybe instead of `x_length` and `y_length` we can introduce the parameters `x_unit` and `y_unit`? It might not seem to make much difference, but it addresses the following points:

4. The default `NumberPlane` has its `x_axis` stretched out a little
```py
plane = NumberPlane()
print(plane.x_axis.unit_size, plane.y_axis.unit_size) 
>>> 1.0158730158730158 1.0
```
As far as I've seen (and used myself), people often use the default `NumberPlane` as a reference to manim's default coordinate system. With the current implementation, users are forced to do `NumberPlane([-config.frame_x_radius, config.frame_x_radius])` to represent the default coord-system. This also forces the use of `plane.c2p(<coords>)` instead of just `coords`. Although it is a good practice to use `plane.c2p`, it seems a bit too much work while working with default coord-system.

5. Setting `unit_size` becomes difficult
If the user wants to set a specific unit size (extremely useful with `LinearTransformation` demos), its tedious with current implementation
```py
x_min, x_max = -4, 4
x_unit = 2
plane = NumberPlane([x_min, x_max], x_length=(x_max - x_min) * x_unit)  
```

6. Visual appeal (also notice something is wrong with the `NumberLine`'s axis ticks:
```py
# master
# plane = NumberPlane(x_min=-5, x_max=6, y_min=-3, y_max=2)
# line = NumberLine(x_min=-config.frame_x_radius, x_max=config.frame_x_radius)
                
# pr branch
plane = NumberPlane([-5, 6], [-3, 2])
line = NumberLine([-config.frame_x_radius, config.frame_x_radius])

line.shift(-line.n2p(0)).shift(DOWN)
line_origin = Dot(line.n2p(0), color=YELLOW)

self.add(plane.set_opacity(0.5), line, line_origin)
```
on master:
![CoordinateSystemsTest_ManimCE_v0 5 0](https://user-images.githubusercontent.com/73361366/115431646-2ac46080-a223-11eb-9df6-4ff88f6c754a.png)
on pr branch:
![CoordinateSystemsTest_ManimCE_v0 5 0](https://user-images.githubusercontent.com/73361366/115431795-521b2d80-a223-11eb-80bd-9ac8fa2eeec0.png)
It doesn't really look good to me that the plane is always stretched out (and with the origin shifted) and you'll have to do extra work setting the `length` as `(max-min)*unit` to set some `unit_size`

Some points might just look like my personal preferences, I'd like to hear others opinions as well :)

hydrobeam replied:
I looked into the issue with the misplaced ticks on the number line and it arises due to the the ticks being drawn starting from the left and continuing onwards. This makes it seem like the dot is misplaced based on tick placement but it's still at 0.

![image](https://user-images.githubusercontent.com/70682032/115930265-26808900-a457-11eb-8a69-c4196e68bab1.png)

As for the odd looking default axis. It would be possible to make it so that x_length and y_length are by default ```x_max-x_min``` so that `unit_size` is always one initially. However, it should be rather intuitive for a user that if `x_range =[-5, 5, 1]` that an x_length of 10 would result in a unit_size of one, or `x_length = 5` would make the unit size two.

abhi3019 commented on manim/mobject/coordinate_systems.py:
```diff
@@ -127,56 +171,55 @@ def input_to_graph_point(self, x, graph):
         axis_config=None,
         x_axis_config=None,
         y_axis_config=None,
-        center_point=ORIGIN,
```
Any specific reason you chose to remove the `center_point` parameter? It's very useful imo

hydrobeam replied:
The changes under `create_axis` make it so that the "`center_point`" is determined dynamically, making it so that defining `center_point` was a bit redundant. Plus, the axes can also be shifted after creation if necessary.


Without `center_point`: 
![image](https://user-images.githubusercontent.com/70682032/115161235-813a6f00-a06a-11eb-86be-90f104c33e60.png)

with the previous default ```center_point = ORIGIN```:
![image](https://user-images.githubusercontent.com/70682032/115161247-9ca57a00-a06a-11eb-94da-731bd3f934f3.png)

abhi3019 replied:
> The changes under create_axis make it so that the "center_point" is determined dynamically

I see, that's nice

> Plus, the axes can also be shifted after creation if necessary

It has to be done like `axes.shift(target_origin - axes.c2p(0, 0))`. Again, this looks like a bit too much work. I'm not against removing the `center_point` parameter, but if the preference is to drop it, let's at least have a method to shift the axes's origin to some other point. Something like
```py
def move_origin_to(self, new_origin, *other_mobs):
    for mob in [self, *other_mobs]:
        mob.shift(new_origin - self.c2p(0, 0))
```
This could probably be done in a later PR, just leaving my thoughts.

hydrobeam replied:
I think this would be better considered in a separate PR.

abhi3019 commented on manim/mobject/coordinate_systems.py:
```diff
@@ -127,56 +171,55 @@ def input_to_graph_point(self, x, graph):
+        new_config["length"] = length
+        axis = NumberLine(range_terms, **new_config)
+
+        # without the if/elif, graph does not exist when min > 0 or max < 0
+        if range_terms[0] > 0:
+            axis.shift(-axis.number_to_point(range_terms[0]))
+        elif range_terms[1] < 0:
+            axis.shift(-axis.number_to_point(range_terms[1]))
+        else:
+            axis.shift(-axis.number_to_point(0))
```
Not really sure what's happening here, but to me it looks like this logic should be implemented in `NumberLine` itself

hydrobeam replied:
I did some light testing and in the current iteration it belongs better in `create_axis`. Shifting this functionality to `NumberLine` would take some work restructuring for little benefit since this is specifcally tuned for `Axes`.

abhi3019 commented on manim/mobject/geometry.py:
```diff
@@ -838,6 +838,21 @@ def get_unit_vector(self):
     def get_angle(self):
         return angle_of_vector(self.get_vector())
 
+    def get_projection(self, point):
```
This is a generally useful function and shouldn't be tied to `Line` imo. Can we move this into `manim\utils\space_ops.py`? It would look like [this](https://github.com/abhi3019/manim-projects/blob/2e831d76ecaa06b37beda00ea0d4c767be24f372/custom_functions/manim.py#L5-L12)

abhi3019 commented on manim/mobject/number_line.py:
```diff
@@ -19,6 +21,235 @@
+            **kwargs,
+        )
+        if self.length:
+            self.rescale_to_fit(self.length, 0, stretch=False)
```
Not a big issue, but we can make use of `Line.set_length` instead. It does the same thing and sounds more intuitive imo
```suggestion
            self.set_length(self.length)
```

hydrobeam commented:
@abhi3019 pointed out that the default `NumberPlane` was stretched so that it did not properly map to the default coordinates. I edited the length values of `NumberPlane` so that it now does this. This does leave the basic `NumberPlane` with some extra space on the sides, but I think it's a worthwhile trade-off. I did consider reverting to the previous method of creating `NumberPlane` background lines (without the box look), but the current implementation looks significantly better for resized planes.

hydrobeam commented:
I've made it so that `NumberPlane` `x_length` and `y_length` scale automatically when not specified to keep the unit size at 1. Plus, the default values for `NumberPlane` are now smaller to accomodate the box look: 

![image](https://user-images.githubusercontent.com/70682032/115945944-1f27a280-a48c-11eb-8b8e-5ced2c9ab346.png)

behackl commented on manim/mobject/number_line.py:
```diff
@@ -19,6 +21,236 @@
 
 
 class NumberLine(Line):
+    def __init__(
```
I'd really like to see a basic docstring for the `NumberLine` class; it doesn't have to include examples -- but a `Parameters`-section where each of the available parameters is briefly explained would be great.

behackl reviewed (changes requested):
I think this PR has been open for long enough, I'd like to get this merged and move on.

I've requested two more changes, but am in favor of merging this afterwards.

GameDungeon commented:
I agree. I LGTM with those changes as well. Thanks a lot Hydrobeam for all the work you put in.

hydrobeam commented:
I've updated the changelog in the PR description and added some documentation for the classes affected by these changes. Let me know if there's anything else that needs to be added.

behackl commented on manim/mobject/number_line.py:
```diff
@@ -19,6 +21,286 @@
+    number_scale_value : :class:`float`
+        The size scaling factor for the number mobjects
+    kwargs : Any
+        Additional arguments to be passed to :class:`~.Line`
```
```suggestion
        Additional arguments to be passed to :class:`~.Line`.
```

behackl commented on manim/mobject/number_line.py:
```diff
@@ -19,6 +21,286 @@
+
+    Returns
+    -------
+    The constructed number line.
```
```suggestion
    NumberLine
        The constructed number line.
```

behackl commented on manim/mobject/number_line.py:
```diff
@@ -19,6 +21,286 @@
+
+
+class NumberLineOld(Line):
+    """
+    The old version of :class:`NumberLine`. Exists for compatibility with GraphScene.
+    """
```
```suggestion
    """The old version of :class:`NumberLine`. 
    
    Exists for compatibility with :class:`~.GraphScene`.
    """
```

behackl reviewed (changes requested):
Thanks for adding documentation! I've looked through it and added a few comments/suggestions, please have a look.

And as a further remark: as soon as we have #1404, the deprecation should be changed to use the new decorators. This can be done in a later PR though.

hydrobeam commented on manim/mobject/coordinate_systems.py:
```diff
@@ -125,58 +169,91 @@ def input_to_graph_point(self, x, graph):
-    def create_axis(self, axis_config):
-        return NumberLine(**merge_dicts_recursively(self.axis_config, axis_config))
+    def create_axis(self, range_terms, axis_config, length):
+        """Creates an axis and dynamically adjusts its position depending on the 0 value.
```
It means 'where 0 is located on the line". I guess it would be clearer to leave it at `adjusts its position`.

behackl commented:
Lets merge this. In case something seriously breaks, we can always do a bugfix release.

Thank you for your efforts -- and your patience. This took a bit longer than I would have liked.

kolibril13 commented:
Nice!
I just had a look on the docs, maybe we can update it here as well:
https://docs.manim.community/en/latest/reference/manim.scene.graph_scene.html?highlight=graph_scene
and here:
https://docs.manim.community/en/latest/examples.html#graphareaplot
there seems to be a broken link
![image](https://user-images.githubusercontent.com/44469195/116743679-8f3ca800-a9f9-11eb-9052-f8323e715703.png)

behackl commented:
> maybe we can update it

What exactly would you like to have updated? The examples?

kolibril13 commented:
````
     y_min=0,
     y_max=100,
     y_axis_config={"tick_frequency": 10},
```
this part
